apm-agent-plugins/apm-micrometer-plugin/src/main/java/co/elastic/apm/agent/micrometer/MicrometerMetricsReporter.java

```diff
@@ -24,81 +24,178 @@
 import co.elastic.apm.agent.matcher.WildcardMatcher;
 import co.elastic.apm.agent.report.Reporter;
 import co.elastic.apm.agent.report.ReporterConfiguration;
+import co.elastic.apm.agent.sdk.logging.Logger;
+import co.elastic.apm.agent.sdk.logging.LoggerFactory;
 import co.elastic.apm.agent.sdk.weakconcurrent.WeakConcurrent;
-import co.elastic.apm.agent.sdk.weakconcurrent.WeakSet;
+import co.elastic.apm.agent.sdk.weakconcurrent.WeakMap;
 import com.dslplatform.json.JsonWriter;
 import io.micrometer.core.instrument.Meter;
 import io.micrometer.core.instrument.MeterRegistry;
 import io.micrometer.core.instrument.composite.CompositeMeterRegistry;
-import co.elastic.apm.agent.sdk.logging.Logger;
-import co.elastic.apm.agent.sdk.logging.LoggerFactory;
+import io.micrometer.core.instrument.simple.CountingMode;
+import io.micrometer.core.instrument.simple.SimpleConfig;
+import io.micrometer.core.instrument.simple.SimpleMeterRegistry;
+import io.micrometer.core.instrument.step.StepRegistryConfig;
 
 import java.io.Closeable;
+import java.time.Duration;
 import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
+import java.util.Set;
 import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.function.Consumer;
 
+/*
+MeterRegistrys with CountingMode CUMULATIVE are eventually accurate, and rarely misleading
+- typically the metric values are immediately correct, but in the worst case are eventually
+correct (ie the metric being written may be a bit behind the last update, but will good enough)
+
+MeterRegistrys with CountingMode STEP update the accessible metric value each "step interval",
+and in between they hold a cache of the last value, and that value is what is provided when
+sampled. So if the agent metric interval is shorter, the agent will publish the cached
+value at least once when it should be producing 0.
+
+Here, the agent takes the step interval as primary if that is being used, otherwise the
+agent interval is used. But note that because we don't know exactly when the meter is
+updated, and the scheduler here can drift long (take more than 1 second between checks),
+it's possible that the agent drops the occasional sample
+ */
 public class MicrometerMetricsReporter implements Runnable, Closeable {
 
+    //interval is actually this plus a random amount of ms more due to processing delays
+    //this means that an occasional step metric will be skipped. This is deemed acceptable
+    private static final long INTERVAL_BETWEEN_CHECKS_IN_MILLISECONDS = 1000L;
```

Comment:
I'm trying to figure if we'd be changing the user-contract - I've set the metrics-interval to 1 second resolution because the alternative of any value (with 1 second minimum) seems to me likely to cause problems. But I haven't though it through. There was a a lot of testing that got me to here, with a lot of different implementations and it still has intermittent failures, so I'm wondering whether this implementation works for now and we can revisit later, or whether I should be retrying with your suggestion now

Reply:
Wasn't the previous user contract that `step_duration < metric_interval` just wouldn't work correctly because it would always skip some steps? So this new restrictions seems fine to me.

To be honest, this all seems quiet complicated without ever having a 100% solution.
Imo, the most maintainable solution would be to request users to always use `CUMULATIVE` due to the micrometer implementation restrictions. This shouldn't be too hard for users (I think?) due to the fact that we request a setup of a `SimpleMeterRegistry` anyway. We could then allow a configuration of which Meters are exposed cumulative vs step on our side (e.g. step_metrics with wildcard matchers like `disable_metrics`) and do the delta calculations safely by ourselves.

I'm not a micrometer expert, so take this all with a big grain of salt and feel free to correct me!

Reply:
A lot of applications already have micrometer meters registered independent of our (or any) agent, and they'll use the stats in their own ways, we're just piggybacking off what's in there already. So I don't think it's valid to insist on `CUMULATIVE` only.

Absolutely agree this is complicated and not ideal

Reply:
Within TSDB, they're planning to create a rate aggregation that can deal with monotonically incrementing counters much better. Maybe it's good enough to document that we recommend `STEP`/delta for now. In the future, `CUMULATIVE`, might be a perfectly valid option, too.